gmp-tcs-offset/src/main/java/edu/gemini/aspen/gmp/tcsoffset/jms/TcsOffsetRequestListener.java

```diff
@@ -0,0 +1,175 @@
+        }
+    }
+
+    /**
+     * Process message sent by client (Instrument)
+     * @param message  Binary client message with the following structure "p q typeOffset"
+     *                 P and Q are the offset sizes. Units arcseconds.
+     *                 typeOffset: Indicate the observation stage Acquisition or Slow Guiding Correciton.
+     */
+
+    private int processApplyOffset(Message message) throws TcsOffsetException {
+        try {
+            Destination replyDestination = message.getJMSReplyTo();
+            BytesMessage msg = (BytesMessage) message;
+
+            // Apply the offset to the TCS
+            sendOffset( msg.readDouble(),
+                        msg.readDouble(),
+                        OffsetType.getFromInt(msg.readInt()),
+                        message.getStringProperty("instName"));
+
+        } catch (JMSException e) {
+            LOG.log(Level.WARNING, "Error reading the msg", e);
+            throw new TcsOffsetException(TcsOffsetException.Error.READING_JMS_MESSAGE,
+                                              "Error reading the JMS message ", e);
+        } catch (TcsOffsetException e) {
+            LOG.log(Level.WARNING, "Problem applying TCS Offset", e);
+            throw e;
+        }
+        return 1;
+    }
+
+    /** 
+     * Check the P and Q offset limits. 
+     * @param obj   : Json Configuration object with the max and min allowd limits
+     * @param value : Value to be checked.
+     */
+
+    private boolean checkLimits(JsonObject obj, double value)  {
+        double max = obj.get("max").getAsDouble();
+        double min = obj.get("min").getAsDouble();
+        LOG.fine("maxC: "+ max + " minC: "+ min + " value: "+ value);
+        if (  Math.abs(value) > Math.abs(max) || Math.abs(value) < Math.abs(min) )
+            return false;
+        return true;
+    }
+
+
+    /**
+     * Check that the offsets are within the allowed limits for each client (Instrument), 
+     * and then send the offset to be applied to the TcsOffsetIOC. 
+     * @param p: P offset value. Units arcsenconds. 
+     * @param q: Q offset value. Units arcsenconds. 
+     * @param instName: Instrument name which want to apply the offset. 
+     * @throws TcsOffsetException: If the offset can be applied send a TcsOffsetException which 
+     *                             has the error code, the Exception trace and the message for the client. 
+     */
+
+    private void sendOffset(double p, double q, OffsetType offsetType, String instName) throws TcsOffsetException {
+        JsonObject obj = _offsetConfig.getAsJsonObject(instName).getAsJsonObject(offsetType.getText()).getAsJsonObject("offset");
+        boolean limitP = checkLimits(obj.getAsJsonObject("p"), p);
+        boolean limitQ = checkLimits(obj.getAsJsonObject("q"), q);
+        LOG.fine("Limit1 : " + limitP + " limit2: " + limitQ);
+        if (!limitP && !limitQ)
```

Comment:
I changed to use this logic during commissioning because sometimes they sent a very small offset for one of the P or Q parameters while the other was correct. So, what I did if one was a disaster, I would apply a 0 to it, and then the correct one would be taken into account.